content/blog/2025-05-06-typed-based-alias-analysis.md

```diff
@@ -0,0 +1,114 @@
+The authors show that their later analyses are within 2.5% of a theoretically perfect analysis when applied to Redundant Load Elimination. One of the main contributions they emphasize is how they also performed a dynamic and limit analysis, in addition to the static analysis.
+
+## Annotations, Rich Types, and Alias Analysis
+
+
+One of the aspects of alias analysis is that it underestimates the true count of aliasing in a program. In TBAA it might be the case that we write a function that may alias but that in the total usages of the function it never actually aliases. In this scenario, we lose the opportunity to optimize some code, but we accept this tradeoff in order to keep the analysis fast and correct. In particular, TBAA is really powerful as it only relies on language constructs to perform the analysis so it doesn’t require special modifications to the source code to improve the results of the alias analysis. But this same approach opens more doors for optimizations if the source language has a richer type system or if there is a way for programmers to provide hints to the compiler. An example is the restrict keyword in C, where the programmers inform the compiler there will be no aliasing between parameters: 
+```C
+//source code
+bool myfn(int* restrict a, int* restrict b) {
+	*a = 1;
+	*b = 2;
+	return *a == *b;
+}
+
+// optimized version
+bool myfn(int* restrict a, int* restrict b) {
+	*a = 1;
+	*b = 2;
+	return false;
+}
+```
+
+In languages with rich type systems the type of the parameter might provide enough information to the compiler so that annotations are not necessary. For example, if we define a function in C++ that receives [unique pointers](https://en.cppreference.com/w/cpp/memory/unique_ptr), we can achieve the same optimization without the need for the use of a restrict keyword: 
+
+```C++
+//source code
+bool myfn(std::unique_ptr<int> a, std::unique_ptr<int> b) {
+	*a = 1;
+	*b = 2;
+	return *a == *b;
+}
+
+// optimized version
+bool myfn(std::unique_ptr<int> a, std::unique_ptr<int> b) {
+	return false;
+}
+```
+In fact, we were even able to optimize assignments to the pointer since we know that they will not be used outside of the scope of the function. The same idea can be used in languages that support immutable types, like [records in Java](https://openjdk.org/jeps/395). 
+
+
+```Java
+//source code
+record MyInt(int value) {}
+boolean myfn(MyInt a, MyInt b) {
+	MyInt a1 = new Record(1)
+ 	MyInt b1 = new Record(2)
+	return a1.value = b1.value
+}
+// optimized version
+boolean myfn(MyInt a, MyInt b) {
+	return false;
+}
+```
+
+So although the original paper did not consider annotations or richer type constructs, the idea can easily be extended to provide even better results for alias analysis. 
```

Comment:
I like this section! It’s great to get somewhat deeper into what language features make TBAA more or less powerful.